modules/webapp-loader/src/main/java/org/wso2/appserver/webapp/loader/AppServerWebappClassLoader.java

```diff
@@ -23,83 +23,83 @@
- * But webapps has the ability to override that behaviour by adding the customised  webapp-classloader.xml file into
+ * Customized WebappClassloader for Application Server. This class introduces a new classloader pattern which defines
+ * on wso2as-web.xml file. The default behaviour is specified in the container level configuration file.
+ * But webapp has the ability to override that behaviour by adding the customised  wso2as-web.xml file into
  * the webapp.
+ *
+ * @since 6.0.0
  */
 public class AppServerWebappClassLoader extends WebappClassLoaderBase {
     private static final Log log = LogFactory.getLog(AppServerWebappClassLoader.class);
 
     private WebappClassLoaderContext webappClassLoaderContext;
 
-    List<URL> urls = new ArrayList<>();
-
     public AppServerWebappClassLoader(ClassLoader parent) {
         super(parent);
     }
 
+    /**
+     * Sets the {@link WebappClassLoaderContext} associated with this classloader
+     * @param classLoaderContext the web application specific classloader context.
+     */
     public synchronized void setWebappClassLoaderContext(WebappClassLoaderContext classLoaderContext) {
-        this.webappClassLoaderContext = classLoaderContext;
+        webappClassLoaderContext = classLoaderContext;
+        webappClassLoaderContext.getProvidedRepositories().forEach(this::addRepository);
     }
 
+    /**
+     * Customized loadClass method which uses {@link WebappClassLoaderContext} to find the classes.
+     * @see ClassLoader#loadClass(String, boolean)
+     */
     @Override
     public synchronized Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundException {
 
-        log.debug("loadClass(" + name + ", " + resolve + ")");
-
-//        Log access to stopped classloader
-//        LifecycleState lifecycleState = this.getState();
-//        if (lifecycleState != LifecycleState.STARTED) {
-//            try {
-//                throw new IllegalStateException();
-//            } catch (IllegalStateException e) {
-//                log.info(sm.getString("webappClassLoader.stopped", name), e);
-//            }
-//        }
+        if (log.isDebugEnabled()) {
+            log.debug("loadClass(" + name + ", " + resolve + ")");
+        }
 
         Class<?> clazz;
 
-        //region Try to load classes from cache
-        // (0) Check our previously loaded local class cache (comes from WebappClassLoaderBase)
+        // Check previously loaded local class cache (comes from WebappClassLoaderBase)
         clazz = super.findLoadedClass0(name);
         if (clazz != null) {
-            log.debug("  Returning class from local cache (WebappClassLoader)");
+            if (log.isDebugEnabled()) {
+                log.debug("  Returning class from local cache (WebappClassLoader)");
+            }
             if (resolve) {
                 resolveClass(clazz);
             }
             return (clazz);
         }
 
```

Comment:
IIRC we discussed to remove this one and test Tomcat context configuration for reading the parent first behaviour. Let me know you did that ?